### PR TITLE
ENH fix rtti issues on osx

### PR DESCRIPTION
**Pull request by beckermr**
Checklist
* [x] Used a [fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

See this issue (https://github.com/conda-forge/clangdev-feedstock/issues/67) for why this PR exists. 

Also, reading the FAQ and ABI info here (https://libcxxabi.llvm.org/) is pertinent to this PR. 

TL;DR This should be safe and helps C++ libraries preserve RTTI information across linkages.

### CODE REVIEW

**Comment by beckermr** [comment]
@conda-forge-admin, please rerender

**Comment by beckermr** [comment]
For all platforms, or just osx?

**Comment by beckermr** [comment]
IIUIC, on linux we still need it.

**Comment by isuruf** [comment]
Let's go with on OSX for now.

**Comment by beckermr** [comment]
Well the tests pass! IDK how the recipe looks. Thoughts?

**Comment by isuruf** [comment]
Can you check that the symbols exported by the new libc++.dylib is the same as the older build number?

**Comment by beckermr** [comment]
This won't build on my mac so I am exporting the symbol list in the ci. I'd like to build it locally on my mac to ensure that the fix works but we might not be able to. Are we saving artifacts in the CI for people to download and test with? That'd be nice right now...

**Comment by beckermr** [comment]
Also the compile tests are not being run.

**Comment by beckermr** [comment]
Alright finally. Here is the diff of the symbols

```
(base) localhost:Desktop Matt$ diff old_syms.txt nw_syms.txt 
2440a2441
> _bcmp
2447d2447
< _chmod
2454a2455
> _fchmodat
```

They are in different spots, but I don't think that matters. I used `nm -gj` to export.

**Comment by beckermr** [comment]
And now conda build is not using the test requirements for outputs like it claims it should. :(

**Comment by beckermr** [comment]
Alright @isuruf. I am filling a ticket with conda-build for the bug in the test env. I am going to comment out those tests for now. 

Here is the diff of the symbols again in case it go lost in the shuffle:

```
(base) localhost:Desktop Matt$ diff old_syms.txt nw_syms.txt 
2440a2441
> _bcmp
2447d2447
< _chmod
2454a2455
> _fchmodat
```

**Comment by beckermr** [comment]
@conda-forge-admin, please rerender

**Comment by isuruf** [comment]
Please don't comment out tests, at least on OSX.

**Comment by beckermr** [comment]
Sure, but do be aware that the tests of the code compiling were not **ever** running. They also won't unless we remove the multiple outputs. Right now conda build won't inject the parent test requirements into the subpackage envs.

**Comment by beckermr** [comment]
issue w/ conda build for cross-reference: https://github.com/conda/conda-build/issues/3777

**Comment by beckermr** [comment]
@conda-forge-admin, please rerender

**Comment by beckermr** [comment]
@conda-forge-admin, please restart build

**Comment by beckermr** [comment]
OK @isuruf. I've solved the issue with the tests (the conda build documentation was wrong) and they all run now. I also built the package locally and it makes the error I reported go away. 

Does the difference in the symbols worry you?

**Comment by beckermr** [comment]
@conda-forge/clangdev Do any of you understand this package conflict for linux?

```
Package libllvm8 conflicts for:
clang -> libllvm8==8.0.1=hc9558a2_0
Package clangdev conflicts for:
clang -> clangdev[version='8.0.0|8.0.1|9.0.0']
Package libgcc-ng conflicts for:
clang -> libgcc-ng[version='>=7.3.0']
Package libcxxabi conflicts for:
libcxx==9.0.0=h14c3975_1 -> libcxxabi==9.0.0=1
Package libstdcxx-ng conflicts for:
clang -> libstdcxx-ng[version='>=7.3.0']
Package gcc_impl_linux-64 conflicts for:
gcc_linux-64 -> gcc_impl_linux-64[version='5.4.0.*|7.2.0.*|7.3.0.*|8.2.0.*']
Package package_has_been_revoked conflicts for:
gcc_linux-64 -> package_has_been_revoked
Package libllvm9 conflicts for:
clang -> libllvm9==9.0.0=hc9558a2_2
Package libcxx conflicts for:
libcxxabi==9.0.0=1 -> libcxx==9.0.0=1
Package clang conflicts for:
clangxx -> clang[version='8.0.0|8.0.1|8.0.1|9.0.0|9.0.0',build='hc9558a2_0|hc9558a2_1|hc9558a2_0|hc9558a2_6|hc9558a2_3|hc9558a2_4|hc9558a2_5|hc9558a2_1']
Package libclang conflicts for:
clang -> libclang==8.0.0
Package binutils_linux-64 conflicts for:
gcc_linux-64 -> binutils_linux-64[version='2.28.1|2.31.1|2.31.1|2.31.1|2.31.1|2.31.1|2.31.1|2.31.1|2.31.1|2.31.1|2.31.1|2.31.1|5.4.0|7.2.0|7.2.0|7.2.0|7.2.0|7.2.0|7.2.0|7.2.0|7.2.0|7.2.0|7.2.0|7.2.0|7.2.0|7.2.0|7.2.0|7.2.0',build='hc67e822_9|hc67e822_15|25|had2808c_24|h6176602_9|h6176602_6|h6176602_2|h6176602_1|h6176602_0|he4fe6c7_1|h6176602_10|h6176602_11|h6176602_12|h6176602_3|h6176602_7|h6176602_8|19|26|had2808c_27|hc67e822_10|hc67e822_11|hc67e822_12|hc67e822_13|hc67e822_14|hc67e822_2|hc67e822_3|hc67e822_7|hc67e822_8']
Package clangxx conflicts for:
clang -> clangxx[version='8.0.0|8.0.1|9.0.0']
Package clang-tools conflicts for:
clang -> clang-tools[version='8.0.0|8.0.1|9.0.0']
```

IDK how or why libllvm8 is getting pulled in.

File: recipe/build.sh [hunk 1]
```diff
@@ -49,11 +58,16 @@ cmake \
...
   -DLLVM_INCLUDE_DOCS=OFF \
   ..
 
 make -j${CPU_COUNT}
 make install
+
+if [[ "$target_platform" == "osx-64" ]]; then
+    # on osx we point libc++ to the system libc++abi
+    install_name_tool -change "@rpath/libc++abi.1.dylib" "/usr/lib/libc++abi.dylib" $PREFIX/lib/libc++.1.0.dylib
+fi
```
Comment (beckermr) [inline recipe/build.sh]: We still need this to adjust where libc++ links. Otherwise when I install the package locally, it fails to find the abi library.

**Comment by isuruf** [comment]
Let's just remove libcxxabi for all platforms.

**Comment by beckermr** [comment]
I am going to build linux against the `libstdc++` ABI provided by the default compilers.

**Comment by beckermr** [comment]
Sure. We are not using it and it is in the git history, but happy to keep all of the windows stuff in the current head.

**Comment by beckermr** [comment]
So building libc++ against the ABI in libstdc++ for linux means that when this library is used, one has to pass `-llibstdc++` at link time even if you intend to use `libc++`. Is this weird @isuruf? IDK who uses this package on linux, but I am guessing this could break things for them. Maybe using `none` on linux might be better?

**Comment by beckermr** [comment]
Th advantage of the build as it stands now is that in principle two libraries, one compiled with libc++ and one compiled with libstdc++, could pass around exceptions or something I think.

**Comment by beckermr** [comment]
meh it is not linking right anyways going to go with none in the interest of finishing this.

File: recipe/build.sh [hunk 2]
```diff
@@ -24,36 +24,37 @@ make -j${CPU_COUNT}
 make install
 cd ..
```
Comment (beckermr) [inline recipe/build.sh]: I think this first build can simple be removed since we don't need it to build cxxabi anymore.

**Comment by beckermr** [comment]
Grrrrrrrrr. We should simply package the cxxabi with the cxx lib and call it good. IDK what the none setting is, but it appears to just not define the symbols it needs. I would have thought `none` would include those symbols in libc++ or something.

**Comment by beckermr** [comment]
There is this option: `LIBCXX_ENABLE_STATIC_ABI_LIBRARY`

**Comment by beckermr** [comment]
Bundling seems fine. We were using a run constrained section with an exact pin.  Thus they were bundled anyways.

**Comment by beckermr** [comment]
OK @conda-forge/libcxx. This is the best I can do. To summarize

1. We have remove the separate libcxxabi library in order to work around some weird solver conflict.
2. On osx, we now pull libcxxabi from the system libraries and we do not ship our own. This solves an exception RTTI problem I was having.
3. On linux, we now bundle libcxx and libcxxabi into a single package called libcxx. The recipe was using exact pins and an exact run constrained, so this procedure is not actually different from what we were doing. 
4. I have reenabled the compilation tests that were being skipped. 
5. I cleaned up some windows stuff we are not using.
6. The ppc64le and aarch64 builds still fail due to a missing dependence (clangdev 9 and llvmdev 9).

Shall we merge?

**Comment by isuruf** [comment]
@beckermr, can you clean up the commits or shall I squash and merge?

**Comment by beckermr** [comment]
squash and merge is great.

**Comment by beckermr** [comment]
Wooooooo! Thank you @isuruf! As usual you are very helpful and I learned a lot!
